구현 초기 생각했던 로직은 다음과 같다.
1. 사용자가 출금 요청
2. 출금 요청이 들어오면, 해당 사용자의 잔액을 확인
3. 잔액이 충분하면 출금 요청 처리
4. 관리자가 출금 요청을 승인
5. 출금 요청이 승인되면, 해당 금액을 사용자 잔액에서 차감

해당 로직을 따라간다면 오류가 없을 것 같았고, 실제로도 이렇게 구현했다.

# 문제점
코드를 다 짠 뒤에, **여러 출금 요청을 동시에 승인**하는 경우를 생각해보게 되었다. <br>
예를 들어, 잔액이 10000원인 사용자가 5000원과 10000원을 각각 출금 요청했다고 가정하자. <br>
사용자는 잔액이 충분하기 때문에 두 출금 요청을 모두 전송할 수 있다. <br>
만약 관리자가 두 요청을 각각 처리한다면(충분한 시간 텀을 두었다고 가정), 문제 없이 두 요청이 승인될 것이다. <br>
하지만, 만약 관리자가 두 요청을 동시에 승인한다면 어떻게 될까? <br>
DB 단에서 생각을 해보자 <br>
> 잔액이 10000원인 지갑
> > 출금 요청 1: 5000원 -> 승인 <br>
> 
> > 출금 요청 2: 10000원 -> 승인

각 트랜잭션이 잔액을 확인했을 때엔 잔액이 충분했기에, commit을 하게 될 것이다. <br>
나는 격리 수준을 따로 설정해두지 않고, @Transactional 어노테이션을 사용했기에 **Reapeatable Read** 격리 수준이 적용된다. <br>
이 격리 수준은 트랜잭션이 시작된 시점의 데이터를 기준으로 읽기 때문에, 두 트랜잭션 모두 잔액이 10000원인 상태에서 출금 요청을 승인하게 된다. <br>
-> 즉 동시성 문제가 발생한다. <br>

# 해결 방법
해결 방법은 크게 두 가지가 있다.
1. **낙관적 락**: 잔액을 업데이트하기 전에 해당 잔액에 대한 버전을 확인한다. 만약 버전이 변경되었다면, 트랜잭션을 롤백하고 다시 시도한다.
2. **비관적 락**: 출금 요청을 승인할 때, 해당 잔액에 대한 락을 걸어 다른 트랜잭션이 접근하지 못하도록 한다. <br>
   이 방법은 데이터베이스에서 제공하는 락을 사용하여, 다른 트랜잭션이 해당 데이터를 수정하지 못하도록 한다.

출금 요청이라는 도메인 특성상 비관적 락을 사용하는 것이 더 적합하다고 판단했는데, <br>
충돌 가능성이 높다고 판단하였고, 낙관적 락을 사용하면 트랜잭션 롤백 + 재시도 로직을 구현해야 하기 때문이다. <br>
따라서, 출금 요청 승인 시 비관적 락을 사용하여 잔액을 업데이트하도록 구현했다. <br>
```java
public interface WalletRepository extends JpaRepository<Wallet, Long> {

    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @Query("SELECT w FROM Wallet w WHERE w.user.id = :userId AND w.currency = :currency")
    Optional<Wallet> findByUserIdAndCurrencyForUpdate(@Param("userId") Long userId, @Param("currency") Wallet.Currency currency);
}

@Transactional
public void processWithdraw(Long userId, BigDecimal amount) {
    Wallet wallet = walletRepository.findByUserIdAndCurrencyForUpdate(userId, Wallet.Currency.KRW)
            .orElseThrow(() -> new WalletException(WALLET_NOT_FOUND));

    wallet.decreaseBalance(amount);
    walletRepository.save(wallet);

    walletLogger(userId, amount);
    notificationService.sendWithdrawNotification(userId, amount);
}
```
이렇게 구현하면, 출금 요청을 승인할 때 해당 지갑 row에 대해 락을 걸어 다른 트랜잭션이 접근하지 못하도록 할 수 있다. <br>
이로 인해, 동시에 여러 출금 요청을 승인하는 경우에도 잔액이 충분한지 확인하고, 잔액을 업데이트하는 과정에서 충돌이 발생하지 않게 된다. <br>
이제 다시 DB 단에서 생각해보자. <br>
> 잔액이 10000원인 지갑
> > 출금 요청 1: 5000원 -> DB Lock, 승인 <br>
>
> > 출금 요청 2: 10000원 -> DB Lock으로 인해 대기 상태 -> DB Lock 해제 -> 잔액이 5000원이므로 승인 실패 <br>

이렇게 비관적 락을 사용하면, 동시에 여러 출금 요청을 승인하는 경우에도 잔액이 충분한지 확인하고, 잔액을 업데이트하는 과정에서 충돌이 발생하지 않게 된다. <br>
동시성 문제는 이렇게 해결할 수 있었고, 이제 다른 문제점도 해결해야 한다. <br>

# Withdraw 테이블
```java
public class Withdraw extends BaseTimeEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne
    private User user;

    private BigDecimal amount;

    @Enumerated(EnumType.STRING)
    private Status status;

    private String bank;

    private String accountNumber;

    @Column(nullable = true)
    private String rejectReason;

    @Column(nullable = true)
    private String failureReason;

    public enum Status {
        PENDING, COMPLETED, REJECTED
    }

    @Builder
    public Withdraw(User user, BigDecimal amount, String bank, String accountNumber, Withdraw.Status status) {
        this.user = user;
        this.amount = amount;
        this.bank = bank;
        this.accountNumber = accountNumber;
        this.status = status;
    }

    public void approve() {
        validatePendingStatus();
        this.status = Withdraw.Status.COMPLETED;
    }

    public void reject(String reason) {
        validatePendingStatus();
        this.status = Withdraw.Status.REJECTED;
        this.rejectReason = reason;
    }

    private void validatePendingStatus() {
        if (this.status != Withdraw.Status.PENDING) {
            throw new WithdrawException(WITHDRAW_STATUS_NOT_PENDING);
        }
    }
}
```
이 테이블은 출금 요청에 대한 정보를 저장하는 테이블이다. <br>
출금 요청이 들어오면, 이 테이블에 출금 요청 정보를 저장하고, 관리자가 출금 요청을 승인하면 상태를 변경한다. <br>
만약 출금 요청을 승인하고, COMPLETED 상태로 변경되었다고 가정하자. <br>
이벤트가 발생하여, 잔액을 업데이트하는 로직이 실행될텐데, 이때 잔액 부족으로 인한 예외가 발생할 수 있다. <br>
이 경우, 출금 요청은 COMPLETED 상태로 변경되었지만, 잔액은 업데이트되지 않은 상태가 된다. <br>
보통, ROLLBACK이 되기 때문에 이런 상황은 발생하지 않겠지만, 나는 이벤트 기반 아키텍쳐를 사용하고 있어 불가능했다. <br>

# 보상 이벤트
이전에, MySQL과 MongoDB를 같이 사용하는 과정에서, 서로 다른 DB 간 트랜잭션이 묶이지 않는다는 것을 알았다. <br>
이때 사용했던 방식이 **보상 트랜잭션**이었다. <br>
단순히 말해 MySQL은 롤백이 되었지만, MongoDB는 롤백이 되지 않는 상황에서 MongoDB의 작업을 취소하는 새로운 트랜잭션을 만들어 실행하는 것이다. <br>
이런 방식으로, 출금 요청이 승인되었지만 잔액 업데이트가 실패한 경우, 출금 요청을 FAILED 상태로 변경하고, 실패 사유를 기록하는 보상 이벤트를 만들면 되지 않을까 싶었다. <br>
우선 엔티티를 수정해주었다.
```java
    @Column(nullable = true)
    private String failureReason;

    public enum Status {
        PENDING, COMPLETED, REJECTED, FAILED
    }
public void fail(String reason) {
    this.status = Withdraw.Status.FAILED;
    this.failureReason = reason;
}
```
status에 FAILED 상태를 추가하고, fail 메서드를 만들어 실패 사유를 기록하도록 했다. <br>
```java
    @RabbitListener(queues = WITHDRAW_APPROVE_QUEUE)
    public void handleWithdrawApproved(WithdrawApprovedEvent event) {
        log.info("출금 승인 이벤트 수신: userId={}, amount={}", event.userId(), event.amount());
        try {
            walletService.processWithdraw(event.userId(), event.amount());
        } catch (WalletException e) {
            log.warn("출금 처리 실패: {}", e.getMessage());
            eventPublisher.publishEvent(new WithdrawFailedEvent(
                    e.getMessage(),
                    event.withdrawId()
            ));
        }
    }
```
이벤트 리스너에서 출금 승인 이벤트를 수신하고, 잔액 업데이트를 시도한다. <br>
잔액 업데이트가 실패하면, WithdrawFailedEvent를 발행하여 보상 이벤트를 처리하도록 한다. <br>
```java
    @RabbitListener(queues = WITHDRAW_FAILURE_QUEUE)
    public void handleWithdrawApprovalFailed(WithdrawFailedEvent event) {
        log.info("출금 실패 이벤트 수신: withdrawId={}, reason={}", event.withdrawId(), event.reason());
        withdrawService.handleWithdrawFailure(event.withdrawId(), event.reason());
    }
```
해당 리스너가 받아 WithdrawFailedEvent를 처리한다. <br>
```java
    @Transactional
    public void handleWithdrawFailure(Long withdrawId, String reason) {
        Withdraw withdraw = withdrawRepository.findById(withdrawId)
                .orElseThrow(() -> new WithdrawException(WITHDRAW_NOT_FOUND));

        withdraw.fail(reason);
        withdrawRepository.save(withdraw);

        notificationSender.send(
                withdraw.getUser().getId(),
                "출금 실패, 사유 : " + reason
        );
    }
```
호출된 메서드가 Withdraw 엔티티를 조회하고, fail 메서드를 호출하여 상태를 FAILED로 변경한다. <br>
그 후, 사용자에게 출금 실패 알림을 전송한다. <br>
이렇게 하면, 출금 요청이 승인되었지만 잔액 업데이트가 실패한 경우에도 출금 요청 상태를 FAILED로 변경하고, 실패 사유를 기록할 수 있다. <br>
이로 인해, 출금 요청과 잔액 업데이트 간의 일관성을 유지할 수 있게 되었다. <br>

# 실제 테스트
![스크린샷 2025-05-25 오전 4 38 02](https://github.com/user-attachments/assets/36addbf8-2d02-4525-9903-5130aa7b1288)
기존 잔액은 10000원이었고, 출금 요청은 5000원과 7000원 두 개가 있었다. <br>
관리자가 두 출금 요청을 동시에 승인했을 때, 첫 요청은 5000원이므로 승인되었고, 두 번째 요청은 잔액이 부족하여 실패 후 이벤트를 발행하였다. <br>
실제 데이터베이스 내부를 확인해보니, 첫 번째 출금 요청은 COMPLETED 상태로 변경되었고, 두 번째 출금 요청은 FAILED 상태로 변경되었다. <br>
![스크린샷 2025-05-25 오전 4 40 14](https://github.com/user-attachments/assets/5eb6070a-dbfd-48bc-88f2-3ca75b1d4e06)<br>
두 번째 출금 요청은, 실패 사유도 기록된 것을 볼 수 있었다.
